brainscore_language/plugins/linear_predictivity/metric.py

```diff
@@ -0,0 +1,180 @@
+        prediction = self._package_prediction(predicted_values, source=source)
+        return prediction
+
+    def _package_prediction(self, predicted_values, source):
+        coords = {coord: (dims, values) for coord, dims, values in walk_coords(source)
+                  if not array_is_element(dims, self._neuroid_dim)}
+        # re-package neuroid coords
+        dims = source.dims
+        # if there is only one neuroid coordinate, it would get discarded and the dimension would be used as coordinate.
+        # to avoid this, we can build the assembly first and then stack on the neuroid dimension.
+        neuroid_level_dim = None
+        if len(self._target_neuroid_values) == 1:  # extract single key: https://stackoverflow.com/a/20145927/2225200
+            (neuroid_level_dim, _), = self._target_neuroid_values.items()
+            dims = [dim if dim != self._neuroid_dim else neuroid_level_dim for dim in dims]
+        for target_coord, target_value in self._target_neuroid_values.items():
+            # this might overwrite values which is okay
+            coords[target_coord] = (neuroid_level_dim or self._neuroid_dim), target_value
+        prediction = NeuroidAssembly(predicted_values, coords=coords, dims=dims)
+        if neuroid_level_dim:
+            prediction = prediction.stack(**{self._neuroid_dim: [neuroid_level_dim]})
+
+        return prediction
+
+    def _align(self, assembly):
+        assert set(assembly.dims) == set(self._expected_dims), \
+            f"Expected {set(self._expected_dims)}, but got {set(assembly.dims)}"
+        return assembly.transpose(*self._expected_dims)
+
+
+class XarrayCorrelation:
+    def __init__(self, correlation, correlation_coord=Defaults.stimulus_coord, neuroid_coord=Defaults.neuroid_coord):
+        self._correlation = correlation
+        self._correlation_coord = correlation_coord
+        self._neuroid_coord = neuroid_coord
+
+    def __call__(self, prediction, target) -> Score:
+        # align
+        prediction = prediction.sortby([self._correlation_coord, self._neuroid_coord])
+        target = target.sortby([self._correlation_coord, self._neuroid_coord])
+        assert np.array(prediction[self._correlation_coord].values == target[self._correlation_coord].values).all()
+        assert np.array(prediction[self._neuroid_coord].values == target[self._neuroid_coord].values).all()
+        # compute correlation per neuroid
+        neuroid_dims = target[self._neuroid_coord].dims
+        assert len(neuroid_dims) == 1
+        correlations = []
+        for i, coord_value in enumerate(target[self._neuroid_coord].values):
+            target_neuroids = target.isel(**{neuroid_dims[0]: i})  # `isel` is about 10x faster than `sel`
+            prediction_neuroids = prediction.isel(**{neuroid_dims[0]: i})
+            r, p = self._correlation(target_neuroids, prediction_neuroids)
+            correlations.append(r)
+        # package
+        result = Score(correlations,
```

Comment:
for the regression code, I find it a bit hard to see how to add in different regression types, hyper parameters, stratification etc., but I really do like this correlation function

Reply:
wondering why these don't subclass `Metric`; are these going to be static core classes that are building blocks of other metrics? what if users want to add their own mappings? it would help to have a common interface

Reply:
AFAIR a couple of weeks ago we discussed that someone else (@aalok-sathe?) would implement the actual fully flexible metric, I added this as more of a reference

Reply:
re `Metric` sub-classing: `CrossRegressedCorrelation` already sub-classed `Metric`, `ScaledCrossRegressedCorrelation` now does too